File: net/JetBrains.Serialization/src/Dataholders/TSTInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using JetBrains.Serialization;

[JsonObject(MemberSerialization.OptIn)]
public class TSTInfo : IEncodableInfo
{
  [JsonProperty("Identifier")] private TextualInfo _identifier;
  [JsonProperty("Content")] private TaggedObjectInfo _content;

  [JsonConstructor]
  public TSTInfo(TextualInfo identifier, TaggedObjectInfo content)
  {
    _identifier = identifier;
    _content = content;
  }

  public TSTInfo(DerSequence sequence)
    : this(
      TextualInfo.GetInstance(sequence[0]),
      new TaggedObjectInfo(
        ((DerTaggedObject)sequence[1]).IsExplicit(),
        ((DerTaggedObject)sequence[1]).TagNo,
        TextualInfo.GetInstance(((DerTaggedObject)sequence[1]).GetObject())))
  {
  }

  public virtual Asn1Encodable ToPrimitive() =>
    new List<IEncodableInfo?>
    {
      _identifier,
      _content
    }.ToPrimitiveDerSequence();
}

File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/dataholders/TSTInfo.kt
package com.jetbrains.signatureverifier.serialization.dataholders

import com.jetbrains.signatureverifier.serialization.toDLSequence
import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.ASN1Primitive
import org.bouncycastle.asn1.DLSequence
import org.bouncycastle.asn1.DLTaggedObject

@Serializable
data class TSTInfo(
  val identifier: TextualInfo,
  val content: TaggedObjectInfo
) : EncodableInfo {

  constructor(sequence: DLSequence) : this(
    TextualInfo.getInstance(sequence.first()),
    sequence.last().let {
      TaggedObjectInfo(
        (it as DLTaggedObject).isExplicit,
        it.tagNo,
        TextualInfo.getInstance(it.baseObject)
      )
    }
  )

  override fun toPrimitive(): ASN1Primitive = listOf(
    identifier.toPrimitive(),
    content.toPrimitive()
  ).toDLSequence()
}